File: app/javascript/models/Card.js
export const suits = ['D', 'H', 'S', 'C']
export const ranks = ['A', 'K', 'Q', 'J', '10', '9', '8', '7', '6', '5', '4', '3', '2']

export default class Card {
    constructor(suit, rank) {
        if (suits.includes(suit) && ranks.includes(rank)) {
            this._suit = suit
            this._rank = rank
        }
    }

    suit() {
        return this._suit
    }

    rank() {
        return this._rank
    }

    equals(card) {
        return (this._suit === card._suit) && (this._rank === card._rank)
    }
}

File: app/models/card.rb
# frozen_string_literal: true 

class Card
  SUITS = ['C', 'H', 'D', 'S']
  RANKS = ['2', '3', '4', '5', '6', '7', '8', '9', '10', 'Jack', 'Queen', 'King', 'Ace']
  HUMANIZED_SUITS = {
    'C' => 'Clubs',
    'H' => 'Hearts',
    'D' => 'Diamonds',
    'S' => 'Spades'
  }

  attr_reader :rank, :suit
  def initialize(rank, suit)
    if RANKS.include?(rank) && SUITS.include?(suit)
      @suit = suit
      @rank = rank
    end
  end

  def ==(other)
    @rank == other.rank && @suit == other.suit
  end

  def <=>(other)
    return 0 unless other
    RANKS.index(@rank) <=> RANKS.index(other.rank)
  end

  def same_rank?(other)
    @rank == other
  end

  def to_s
    "#{rank} of #{HUMANIZED_SUITS.fetch(suit)}"
  end

  def as_json
    {
      rank: rank,
      suit: suit,
    }
  end

  def self.from_json(json)
    self.new(json[:rank], json[:suit])
  end
end